### PR TITLE
Fix race with launch context changes when loading composable nodes

### PR DESCRIPTION
**Pull request by jacobperron**
This bug was discovered when trying load composable nodes from a GroupAction.
The ROS namespace (and presumably other remaps)  pushed onto the context stack was popped after the LoadComposableNodes execute() function finished.
But because the loading happens asynchronously, we need to make sure we get the necessary information from the context before execute() finishes.

### CODE REVIEW

**Comment by ivanpauno** [comment]
@jacobperron Does this fix https://github.com/ros2/launch_ros/issues/114?

**Comment by jacobperron** [comment]
> Can you add a regression test?

Will do.

> Does this fix #114?

I had completely forgotten about that issue. I'll check to see if it fixes it.

**Comment by jacobperron** [comment]
So, it appears this PR fixes the example mentioned in https://github.com/ros2/launch_ros/issues/114#issuecomment-624904000, which involves setting the namespace, but not the original issue that involves setting parameters on the node. I'll look into if a similar fix to the one proposed in this PR can be used to fix #114.

**Comment by ivanpauno** [comment]
> So, it appears this PR fixes the example mentioned in [#114 (comment)](https://github.com/ros2/launch_ros/issues/114#issuecomment-624904000), which involves setting the namespace, but not the original issue that involves setting parameters on the node. I'll look into if a similar fix to the one proposed in this PR can be used to fix #114.

That's weird, because `get_composable_node_load_request` also resolves the parameters to be loaded.

**Comment by jacobperron** [comment]
> That's weird, because get_composable_node_load_request also resolves the parameters to be loaded.

Is there a difference between "global" params (aka launch config `ros_params`) and parameters passed directly to a node?

**Comment by jacobperron** [comment]
hmm, but it's not the parameters that are being dropped, but specifically a user-defined launch configuration "bar_arg". I can do some more debugging.

**Comment by jacobperron** [comment]
I've added some regression tests in 719ced8.

There's still a bug related to parameters (and some other corner-cases); I'll open up separate PRs for those.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        namespace=namespace,
+        output='screen',
+        parameters=parameters,
+        remappings=remappings,
+    )
+
+
+def _load_composable_node(*, parameters=None, remappings=None):
+    return LoadComposableNodes(
+        target_container=TEST_CONTAINER_NAME,
```
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:64]: what about the container namespace?
`_create_node_container` is optionally using one.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        composable_node_descriptions=[
+            ComposableNode(
+                package='nonexistent_package',
+                plugin='this_plugin_should_not_exist',
+                name=TEST_NODE_NAME,
+            )
+        ])
+    context = _assert_launch_no_errors([_create_node_container(), load_composable_nodes])
+
+    assert get_node_name_count(context, f'/{TEST_NODE_NAME}') == 0
```
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:89]: should this be tested here?
sounds unrelated to me.

Maybe, what you want is checking the value of the [response](https://github.com/ros2/launch_ros/blob/9250313932c2c7477a982ef0c0cfe5124d3cfddf/launch_ros/launch_ros/actions/load_composable_nodes.py#L116). That could be introspected if the action would store the responses.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            ],
+            scoped=True,
+        ),
+    ])
+
+    assert get_node_name_count(context, f'/{TEST_NODE_NAME}') == 1
+    assert get_node_name_count(context, f'/{TEST_CONTAINER_NAME}') == 1
+
+    # Check the remaps in load service request
+    assert len(load_composable_nodes_action._LoadComposableNodes__load_node_requests) == 1
```
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:121]: maybe the sent request could be a property, so we don't have to check a private member variable

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 2 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:89]: It could actually be checked in the request instead of the response

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 3 above)
Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:121]: I followed a similar pattern that is used in [test_node.py](https://github.com/ros2/launch_ros/blob/9250313932c2c7477a982ef0c0cfe5124d3cfddf/test_launch_ros/test/test_launch_ros/actions/test_node.py#L91).

I think a user should have access to that information; it seems like an implementation detail subject to change.

Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:121]: > I followed a similar pattern that is used in test_node.py.

Yeah, it's not a great pattern, but acceptable I guess.

> I think a user should have access to that information; it seems like an implementation detail subject to change.

That sounds contradictory, I guess you meant "should not".

---

I prefer not testing implementation details, but I don't feel strongly in this case.

Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:121]: Yeah, sorry, I meant "should not" (updated my comment).

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 2 above)
Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:89]: I wanted to check that the node is not launched. I think this is a proxy for that; maybe there's a better way. The request  contains the same information we passed in the composable node description, regardless if the loading was successful or not.

We *could* check the response too, but I was avoiding the storing more information than was necessary. I can add it if you think it would be better.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 1 above)
Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py]: Good point, I've made the target container a parameter in a12906b

We're not using the namespace in any tests, but I might use of it in follow-up patches.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 3 above)
Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:126]: I guess the ideal way to write tests in this situation would be to write a mock component container and assert things about it (e.g. if the correct service request was received). It adds a bit more to the test complexity, but could be better long-term. What do you think?

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 1 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py]: thanks

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 2 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:94]: Ok, I see what this is testing now.
The approach is a bit weird, but it seems acceptable.

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py (hunk 3 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:126]: > What do you think?

Sounds like a good proposal to me, though I don't mind approving this as-is.

Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py]: I went ahead and refactored the tests to use a mock object. I think it's much nicer. PTAL 5d9dcf3

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    assert request.plugin_name == 'bar_plugin'
+    assert request.node_name == 'test_node_name'
+    assert request.node_namespace == '/test_node_namespace'
+    assert len(request.remap_rules) == 0
+    assert len(request.parameters) == 2
+    assert request.parameters[0].name == 'test_param1'
+    assert request.parameters[0].value.type == ParameterType.PARAMETER_STRING
+    assert request.parameters[0].value.string_value == 'test_value_param1'
+    assert request.parameters[1].name == 'test.param2'
+    # TODO(jacobperron): I would expect this to be a double value, but we can only pass strings
```
Comment (jacobperron) [inline test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py:220]: I'll take a look at resolving this in a follow-up.

**Comment by jacobperron** [comment]
ugh, there's an issue with the tests where they hang sometimes... I can reproduce locally. I think there's an issue shutting down the launch context *before* LoadComposableNodes has had a chance to make a service request.

**Comment by jacobperron** [comment]
I've discovered a race in launch (reported in https://github.com/ros2/launch_ros/issues/169), with a proposed fix in https://github.com/ros2/launch/pull/449. With the proposed fix, the tests in the PR pass, otherwise we can expect them to hang forever.

**Comment by ivanpauno** [comment]
> I've discovered a race in launch (reported in #169), with a proposed fix in [ros2/launch#449](https://github.com/ros2/launch/pull/449). With the proposed fix, the tests in the PR pass, otherwise we can expect them to hang forever.

I understand the race, but what is causing it here?
The thread you're creating isn't emiting any event, so I would like to understand what exactly is causing the race here (the launch service is running in a single threaded event loop)

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11934/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6977/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9724/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11873/)

**Comment by ivanpauno** [comment]
@jacobperron I've cancelled a couple of jobs that were hanging

**Comment by jacobperron** [comment]
Sadly, it still seems like there's a race somewhere causing the tests to hang somteimes :disappointed:

**Comment by jacobperron** [comment]
I wasn't properly shutting down the mock container and so the synchronous service call in `LoadComposableNodes` was blocking indefinitely. I've fixed the test in 9c50db2 and opened up a follow-up ticket to improve `LoadComposableNodes` such that we avoid this situation https://github.com/ros2/launch_ros/issues/171.

**Comment by jacobperron** [comment]
Note, I don't expect the PR CI job to pass until ros2/launch#449 is released.

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11948/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6991/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9740/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11891/)

File: test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py [hunk 5]
```diff
@@ -144,6 +146,7 @@ def test_load_node(mock_component_container):
     assert len(request.extra_arguments) == 0
 
 
+
```
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py]: nit: too much spacing(?)

Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_load_composable_nodes.py]: Fixed in cf1437a

**Comment by jacobperron** [comment]
@ros-pull-request-builder retest this please
